### ISSUE TITLE
decryptSync() Marks Decrypted Fields As Modified

### ISSUE THREAD

**Issue by dereekb** [joegoldbeck/mongoose-encryption#100]
Ran into an issue today where calling decryptSync() will mark all values that were decrypted as modified, or rather mongoose marks the fields as modified when the fields are assigned to.

I'm not sure this is desired behavior or not. My interpretation is that the fields should not be marked modified right after decryption since we'd want to track if those fields change or not.

Since I'm using a custom decrypt plugin (that calls decryptSync()) I mitigated this issue by just setting all modified paths to be marked unmodified.

```
    // Unmark all paths as modified.
    doc.modifiedPaths().forEach((path) => {
      doc.unmarkModified(path);
    });
```

**Comment by joegoldbeck** [comment]
This is certainly not desired behavior the way you describe it. To be clear, you are calling `decryptSync` and then not modifying any values at all, and yet they are marked modified?

**Comment by dereekb** [comment]
Yep, that's what I'm experiencing. Should be easy to see with a test.

I did double check to make sure nothing else was calling it, or that it wasn't already modified prior to being decrypted, but these models were just loaded out of the database.

If you don't find the same behavior, then it may just be on my end.

**Comment by joegoldbeck** [comment]
I confirmed this behavior on the branch https://github.com/joegoldbeck/mongoose-encryption/tree/marked-modified

Tests are there, though they seem to be a little wonky (the first one with `decrypt` fails, but the second one with `decryptSync` passes, and the first one failed even if I change it to `decryptSync`...). Not quite sure what's going on there.

Also not totally sure what the right fix is here. Within the `decryptSync` method, the document seems to be a vanilla js object, so I can't `unmarkModified` things 📦 

Feel free to branch off and take a swing if you're interested.

**Comment by joegoldbeck** [comment]
One concern here is that it's possible to both modify those fields _and_ then decrypt the document. I don't think we can just assume that the decrypted document has had no modifications to those fields.

**Comment by dereekb** [comment]
Is it valid to decrypt an encrypted document and the decrypted values (values that are not excluded from encryption/decryption) persist as modified?

My interpretation about desired behavior is that if it isn't marked modified then decryption would/could theoretically overwrite those values, or is that not correct?

Is it valid to have a document with _ct set, change some values around, then decrypt and the results be merged? I think we do know what fields are decrypted and were set on _ct, so it should be ok to mark those as not modified.

For my case I'm only decrypting after save and don't mind blowing away the isModified state. I agree though that the values that aren't decrypted should retain any isModified states. We'll have to make sure that tests cover those cases properly.

I don't have time in the immediate to tackle it, but I'll try and come back around and contribute when I get a chance. For now just wanted to make the issue known incase someone else ran into it!

**Comment by joegoldbeck** [comment]
> My interpretation about desired behavior is that if it isn't marked modified then decryption would/could theoretically overwrite those values, or is that not correct

I think that is correct. I could imagine a situation where values were modified on a decrypted document, then it gets encrypted, then it gets decrypted. It's preeeeetty edge-casy.

I _think_ the desired behavior would be to preserve the isModified state across the decryption, whatever it was, since decryption shouldn't change that. Seems probably safer to have something marked modified when it wasn't than have it be unmarked when it actually was. Then again, maybe the modified status of these encrypted fields doesn't really matter all that much in most case since they never get saved to Mongo themselves (only as _ct)?

**Comment by joegoldbeck** [comment]
Thanks for pointing it out @dereekb! I think leaving this conversation as a placeholder in case someone else runs into it is a good solution for now :)
